teos/teosd.py

```diff
@@ -68,37 +61,76 @@ def get_config(command_line_conf, data_dir):
+        watcher (:obj:`teos.watcher.Watcher`) the `Watcher` instance.
+        chain_monitor (:obj:`teos.chain_monitor.ChainMonitor`) the `ChainMonitor` instance.
+        self.api_popen (:obj:`subprocess.Popen` or "obj"`None`) if set, the `Popen` instance running the public API.
+        self.api_process (:obj:`multiprocessing.Process` or "obj"`None`) if set, the `Process` instance running the public API.
+        self.rpc_process (:obj:`multiprocessing.Process`) the instance of the internal RPC server; only set if running.
+        self.internal_api (:obj:`teos.internal_api.InternalAPI`) the InternalAPI instance.
+    """
+
+    def __init__(self, config):
+        self.config = config
+
+        # event triggered when a `stop` command is issued
+        # Using multiprocessing.Event seems to cause a deadlock if event.set() is called in a signal handler that
+        # interrupted event.wait(). This does not happen with threading.Event.
+        # See https://bugs.python.org/issue41606
+        self.stop_command_event = threading.Event()
+
+        # event triggered when the public API is halted, hence teosd is ready to stop
+        self.stop_event = multiprocessing.Event()
 
-        setup_data_folder(config.get("DATA_DIR"))
-        setup_logging(config.get("LOG_FILE"))
+    def start(self):
+        """This method implements the whole lifetime cycle of the the TEOS tower. This method does not return."""
+        try:
+            logger.info("Starting TEOS")
+            self.setup_components()
+            self.start_services()
 
-        logger.info("Starting TEOS")
+            self.stop_command_event.wait()
 
-        bitcoind_connect_params = {k: v for k, v in config.items() if k.startswith("BTC_RPC")}
-        bitcoind_feed_params = {k: v for k, v in config.items() if k.startswith("BTC_FEED")}
+            self.teardown()
+
+        except Exception as e:
+            logger.error("An error occurred: {}. Shutting down".format(e))
+            exit(1)
+
+    def setup_components(self):
+        """Performs the initial setup and creates all the components."""
+        setup_data_folder(self.config.get("DATA_DIR"))
+        setup_logging(self.config.get("LOG_FILE"))
+
+        bitcoind_connect_params = {k: v for k, v in self.config.items() if k.startswith("BTC_RPC")}
+        bitcoind_feed_params = {k: v for k, v in self.config.items() if k.startswith("BTC_FEED")}
 
         if not can_connect_to_bitcoind(bitcoind_connect_params):
             logger.error("Cannot connect to bitcoind. Shutting down")
 
-        elif not in_correct_network(bitcoind_connect_params, config.get("BTC_NETWORK")):
-            logger.error("bitcoind is running on a different network, check conf.py and bitcoin.conf. Shutting down")
+        elif not in_correct_network(bitcoind_connect_params, self.config.get("BTC_NETWORK")):
+            logger.error("bitcoind is running on a different network, check teos.conf and bitcoin.conf. Shutting down")
```

Comment:
Done in 590f7556a0bb1e9c392fd5a7b4d210cdacc4958f. I didn't find a more specific error rather than RuntimeError, but I guess that's fine. (same for other similar error below).